src/backend/gl/src/command.rs

```diff
@@ -235,17 +244,94 @@ impl From<hal::Limits> for Limits {
-/// `display_fb` field.
 #[derive(Debug)]
-pub struct CommandBuffer {
+pub struct BufferStorage {
     pub(crate) memory: Arc<Mutex<BufferMemory>>,
     pub(crate) buf: BufferSlice,
     // Buffer id for the owning command pool.
     // Only relevant if individual resets are allowed.
     pub(crate) id: u64,
+}
+
+impl BufferStorage {
+    fn push_cmd(&mut self, cmd: Command) {
+        let mut memory = self
+            .memory
+            .try_lock()
+            .expect("Trying to record a command buffers, while memory is in-use.");
+
+        let cmd_buffer = match *memory {
+            BufferMemory::Linear(ref mut buffer) => &mut buffer.commands,
+            BufferMemory::Individual {
+                ref mut storage, ..
+            } => &mut storage.get_mut(&self.id).unwrap().commands,
+        };
+
+        cmd_buffer.push(cmd);
+
+        self.buf.append(BufferSlice {
+            offset: cmd_buffer.len() as u32 - 1,
+            size: 1,
+        });
+    }
+
+    /// Copy a given vector slice into the data buffer.
+    fn add<T>(&mut self, data: &[T]) -> BufferSlice {
+        self.add_raw(unsafe {
+            slice::from_raw_parts(data.as_ptr() as *const _, data.len() * mem::size_of::<T>())
+        })
+    }
+
+    /// Copy a given u8 slice into the data buffer.
+    fn add_raw(&mut self, data: &[u8]) -> BufferSlice {
+        let mut memory = self
+            .memory
+            .try_lock()
+            .expect("Trying to record a command buffers, while memory is in-use.");
+
+        let data_buffer = match *memory {
+            BufferMemory::Linear(ref mut buffer) => &mut buffer.data,
+            BufferMemory::Individual {
+                ref mut storage, ..
+            } => &mut storage.get_mut(&self.id).unwrap().data,
+        };
+        data_buffer.extend_from_slice(data);
+        let slice = BufferSlice {
+            offset: (data_buffer.len() - data.len()) as u32,
+            size: data.len() as u32,
+        };
+        slice
+    }
+
+    fn reset(&mut self) {
+        let mut memory = self
+            .memory
+            .try_lock()
+            .expect("Trying to reset a command buffer, while memory is in-use.");
+
+        match *memory {
+            // Linear` can't have individual reset ability.
+            BufferMemory::Linear(_) => unreachable!(),
+            BufferMemory::Individual {
+                ref mut storage, ..
+            } => {
+                // TODO: should use the `release_resources` and shrink the buffers?
+                storage.get_mut(&self.id).map(|buffer| {
+                    buffer.commands.clear();
+                    buffer.data.clear();
+                });
+            }
+        }
+    }
+}
+
+/// A command buffer abstraction for OpenGL.
+///
+/// If you want to display your rendered results to a framebuffer created externally, see the
+/// `display_fb` field.
+#[derive(Debug)]
+pub struct CommandBuffer {
+    pub(crate) buffer: BufferStorage,
```

Comment:
maybe something like `data: CommandStorage` or something?

Reply:
sure, I was a bit undecisive of name. Will make it `CommandStorage`